### PR TITLE
Properly handle the required 'step' in optimizer states

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
Properly handle the required 'step' in optimizer states in higher version torch.

### CODE REVIEW

File: alf/optimizers/optimizers.py [hunk 1]
```diff
@@ -501,6 +507,28 @@ def load_state_dict(self, state_dict):
...
+
+        for param_group in self.param_groups:
+            for p in param_group['params']:
+                p_state = self.state[p]
+                if 'step' not in p_state:
+                    # optimizers in higher version torch requires 1) the presence of 'step'
+                    # in state, and 2) it appears as the first key of the state dictionary.
+                    # Therefore we explicitly create it if it does not exist and make it
+                    # the key that comes first
+                    p_state.update({'step': 0})
```
Comment (emailweixu) [inline alf/optimizers/optimizers.py]: Is this needed by all optimizers?

Comment (Haichao-Zhang) [reply alf/optimizers/optimizers.py]: Some of the optimizers have this requirement, such Adam, AdamW, ASGD. Here we handle it in a general way without differentiating different variations while without too much additional overhead. Updated comment

File: alf/optimizers/optimizers.py [hunk 2]
```diff
@@ -501,6 +507,28 @@ def load_state_dict(self, state_dict):
...
+                    # Some of the optimizers such as Adam/AdamW/ASGD etc in higher version
+                    # torch require 1) the presence of 'step' in state, and 2) it appears
+                    # as the first key of the state dictionary. Therefore we explicitly
+                    # create it if it does not exist and make it the key that comes first
+                    p_state.update({'step': 0})
+                    p_state = {
+                        **_move_key_as_the_first(p_state, 'step'),
+                        **p_state
+                    }
+                    self.state[p] = p_state
```
Comment (emailweixu) [inline alf/optimizers/optimizers.py]: can be simplified as: `self.state[p] = {'step': 0, **p_state}`

Comment (Haichao-Zhang) [reply alf/optimizers/optimizers.py]: Simplified. Thanks for the suggestion.

**Review by emailweixu — APPROVED** [review]

**Pull request merged**